File: src/CSharp/653.TwoSumBST.cs

using System;
using System.Collections.Generic;
using System.Text;

namespace PeefyLeetCode.TwoSumBST
{
    public class TreeNode
    {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int x) { val = x; }
    }

    public class Solution
    {
        HashSet<int> s = new HashSet<int>();
        public bool FindTarget(TreeNode root, int k)
        {
            if (root == null)
                return false;
            if (s.Contains(root.val))
                return true;
            else
                s.Add(k - root.val);
            return FindTarget(root.left, k) || FindTarget(root.right, k);
        }
    }
}

File: src/Java/653.TwoSumBST.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    public boolean findTarget(TreeNode root, int k) {
        if (root == null)
            return false;
        LinkedList<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        HashSet<Integer> nums = new HashSet<>();
        while (queue.size() > 0){
            TreeNode node = queue.pollFirst();
            if (nums.contains(k - node.val))
                return true;
            nums.add(node.val);
            if (node.left != null)
                queue.add(node.left);
            if (node.right != null)
                queue.add(node.right);
        }
        return false;
    }
}

